hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -30,125 +30,189 @@
-                i += 1
-
-    run.command = description
-    run.__name__ = "block_program(%r)" % (description,)
-    return run
-
-
-class PassClassification(Enum):
-    CANDIDATE = 0
-    HOPEFUL = 1
-    DUBIOUS = 2
-    AVOID = 3
-    SPECIAL = 4
-
-
-@total_ordering
-@attr.s(slots=True, cmp=False)
-class ShrinkPass(object):
-    pass_function = attr.ib()
-    index = attr.ib()
-
-    classification = attr.ib(default=PassClassification.CANDIDATE)
-
-    successes = attr.ib(default=0)
-    runs = attr.ib(default=0)
-    calls = attr.ib(default=0)
-    shrinks = attr.ib(default=0)
-    deletions = attr.ib(default=0)
-
-    @property
-    def failures(self):
-        return self.runs - self.successes
-
-    @property
-    def name(self):
-        return self.pass_function.__name__
-
-    def __eq__(self, other):
-        return self.index == other.index
-
-    def __hash__(self):
-        return hash(self.index)
-
-    def __lt__(self, other):
-        return self.key() < other.key()
-
-    def key(self):
-        # Smaller is better.
-        return (self.runs, self.failures, self.calls, self.index)
+    return (len(buffer), buffer)
 
 
 class Shrinker(object):
     """A shrinker is a child object of a ConjectureRunner which is designed to
-    manage the associated state of a particular shrink problem.
+    manage the associated state of a particular shrink problem. That is, we
+    have some initial ConjectureData object and some property of interest
+    that it satisfies, and we want to find a ConjectureData object with a
+    shortlex (see sort_key above) smaller buffer that exhibits the same
+    property.
+
+    Currently the only property of interest we use is that the status is
+    INTERESTING and the interesting_origin takes on some fixed value, but we
+    may potentially be interested in other use cases later.
+    However we assume that data with a status < VALID never satisfies the predicate.
+
+    The shrinker keeps track of a value shrink_target which represents the
+    current best known ConjectureData object satisfying the predicate.
+    It refines this value by repeatedly running *shrink passes*, which are
+    methods that perform a series of transformations to the current shrink_target
+    and evaluate the underlying test function to find new ConjectureData
+    objects. If any of these satisfy the predicate, the shrink_target
+    is updated automatically. Shrinking runs until no shrink pass can
+    improve the shrink_target, at which point it stops. It may also be
+    terminated if the underlying engine throws RunIsComplete, but that
+    is handled by the calling code rather than the Shrinker.
+
+    =======================
+    Designing Shrink Passes
+    =======================
+
+    Generally a shrink pass is just any function that calls
+    cached_test_function and/or incorporate_new_buffer a number of times,
+    but there are a couple of useful things to bear in mind.
+
```

Comment:
You mention it in passing in the example below, but it's probably worth having a dedicated paragraph to explain that shrink passes need to be robust against mostly-arbitrary structural changes to examples/blocks that can occur whenever a test call succeeds, even in the middle of a shrink loop.